ffi/capi/src/neo_datetime.rs

```diff
@@ -1012,33 +1606,180 @@ pub mod ffi {
+    pub struct NeoZonedDateTimeFormatterGregorian(
+        pub  icu_datetime::FixedCalendarDateTimeFormatter<
+            Gregorian,
+            icu_datetime::fieldsets::enums::CompositeFieldSet,
+        >,
+    );
+
+    impl NeoZonedDateTimeFormatterGregorian {
+        #[diplomat::rust_link(icu::datetime::FixedCalendarDateTimeFormatter::format, FnInStruct)]
+        #[diplomat::rust_link(icu::datetime::FormattedDateTime, Struct, hidden)]
+        #[diplomat::rust_link(icu::datetime::FormattedDateTime::to_string, FnInStruct, hidden)]
+        pub fn format_iso(
+            &self,
+            date: &IsoDate,
+            time: &Time,
+            zone: &TimeZoneInfo,
+            write: &mut diplomat_runtime::DiplomatWrite,
+        ) -> Result<(), DateTimeWriteError> {
+            let date = date.0.to_calendar(Gregorian);
+            let mut input = icu_datetime::DateTimeInputUnchecked::default();
+            input.set_date_fields(date);
+            input.set_time_fields(time.0);
+            input.set_time_zone_id(zone.time_zone_id);
+            if let Some(offset) = zone.offset {
+                input.set_time_zone_utc_offset(offset);
+            }
+            if let Some(local_time) = zone.local_time {
+                input.set_time_zone_local_time(local_time);
+            }
+            if let Some(zone_variant) = zone.zone_variant {
+                input.set_time_zone_variant(zone_variant);
+            }
+            let _infallible = self
+                .0
+                .format_unchecked(input)
+                .try_write_to(write)
+                .ok()
+                .transpose()?;
+            Ok(())
         }
     }
 }
 
-#[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
-fn map_or_default<Input, Output>(input: Option<Input>) -> Output
-where
-    Output: From<Input> + Default,
-{
-    input.map(Output::from).unwrap_or_default()
+mod impls {
+    #[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+    use alloc::boxed::Box;
+    #[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+    use icu_calendar::Gregorian;
+    #[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+    use icu_datetime::{
+        fieldsets::enums::*, fieldsets::Combo, pattern::*, scaffold::*, DateTimeFormatter,
+        DateTimeFormatterLoadError, FixedCalendarDateTimeFormatter,
+    };
+
+    #[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+    pub(super) fn map_or_default<Input, Output>(input: Option<Input>) -> Output
+    where
+        Output: From<Input> + Default,
+    {
+        input.map(Output::from).unwrap_or_default()
+    }
+
+    #[cfg(any(feature = "compiled_data", feature = "buffer_provider"))]
+    pub(super) fn formatter_with_zone<Zone>(
+        formatter: &DateTimeFormatter<CompositeDateTimeFieldSet>,
+        locale: &crate::locale_core::ffi::Locale,
+        zone: Zone,
+        load: impl FnOnce(
```

Comment:
issue: this signature is a mess: can we use some clever imports to make it cleaner? Fine with things being renamed to FfiLocale/etc

Reply:
This is code I pulled out of the individual constructors above, and these are what the signatures end up being. These functions are extremely private and internal. I think having a private function with messy signatures is better than having this be either macro-generated or template-generated, which was the alternative I considered.

Reply:
@sffc I'm not suggesting rearchitecting this, I'm suggesting flattening the imports. The signature has inherent complexity _and_ line noise from the fully qualified paths, each of these on its own is not bad but put together it gets really hard to read. I'm suggesting we fix the latter problem, which doesn't require any rearchitecture.

Reply:
looks like you mostly did that in your latest commit, so I'm happy. I'd also fix the `Locale` import but that's not a huge deal.

Reply:
Yeah I had already imported most things but there were two more that I fixed. Same signature, just not as many characters in code.

Reply:
So regarding my strategy for imports: mostly in `mod ffi` we import types from the icu_capi crate. I made this `mod impls` so that I would import types from the Rust crates. The names have a high probability of clashing, so I prefer to keep modules importing only one or the other flavor.